src/main/java/org/whispersystems/textsecuregcm/storage/DirectoryReconciler.java

```diff
@@ -0,0 +1,248 @@
+                             DirectoryReconciliationCache reconciliationCache,
+                             AccountsManager accountsManager) {
+    this.accountsManager      = accountsManager;
+    this.reconciliationClient = reconciliationClient;
+    this.reconciliationCache  = reconciliationCache;
+    this.random               = new SecureRandom();
+    this.workerId             = generateWorkerId(random);
+  }
+
+  private static String generateWorkerId(SecureRandom random) {
+    byte[] workerIdBytes = new byte[16];
+    random.nextBytes(workerIdBytes);
+    return Hex.toString(workerIdBytes);
+  }
+
+  @Override
+  public void start() {
+    start(Executors.newSingleThreadScheduledExecutor());
+  }
+
+  @VisibleForTesting
+  public void start(ScheduledExecutorService executor) {
+    this.executor = executor;
+
+    try {
+      reconciliationCache.cleanUpWorkerSet(WORKER_TTL_MS);
+    } catch (Exception ex) {
+      logger.warn("failed to clean up worker set", ex);
+    }
+
+    this.executor.scheduleWithFixedDelay(new PingTask(workerId, reconciliationCache), 0, WORKER_TTL_MS / 2, TimeUnit.MILLISECONDS);
+    scheduleWithJitter(CHUNK_INTERVAL);
+  }
+
+  @Override
+  public void stop() {
+    this.executor.shutdown();
+
+    try {
+      this.executor.awaitTermination(5000, TimeUnit.MILLISECONDS);
+    } catch (InterruptedException ex) {
+      throw new AssertionError(ex);
+    }
+
+    try {
+      reconciliationCache.leaveWorkerSet(workerId);
+    } catch (Exception ex) {
+      logger.warn("failed to remove from worker set", ex);
+    }
+  }
+
+  @Override
+  public void run() {
+    boolean accelerate  = false;
+    long    workerCount = 1;
+
+    try {
+      if (reconciliationCache.lockActiveWorker(workerId, CHUNK_INTERVAL)) {
+        try {
+          if (processChunk()) {
+            accelerate = reconciliationCache.isAccelerated();
+          }
+        } finally {
+          reconciliationCache.unlockActiveWorker(workerId);
+        }
+      }
+
+      if (!accelerate) {
+        workerCount = Math.max(reconciliationCache.getWorkerCount(WORKER_TTL_MS), 1);
+      }
+    } catch (Exception ex) {
+      logger.warn("error in directory reconciliation", ex);
+    }
+
+    if (!accelerate) {
+      scheduleWithJitter(CHUNK_INTERVAL * workerCount);
+    } else {
+      scheduleWithJitter(ACCELERATE_CHUNK_INTERVAL);
+    }
+  }
+
+  private void scheduleWithJitter(long delayMs) {
+    long randomJitterMs = (long) (random.nextDouble() * JITTER_MAX * delayMs);
+    try {
+      executor.schedule(this, delayMs + randomJitterMs, TimeUnit.MILLISECONDS);
+    } catch (RejectedExecutionException ex) {
+      logger.info("reconciler shutting down");
+    }
+  }
+
+  private boolean processChunk() {
+    Optional<String>               fromNumber          = reconciliationCache.getLastNumber();
+    int                            chunkSize           = (int) Math.max(getAccountCount() * PERIOD / CHUNK_INTERVAL, MINIMUM_CHUNK_SIZE);
```

Comment:
I'm curious why the chunk size is dynamic. Shouldn't the maximum reasonable chunk size to create a query that doesn't time out be constant, regardless of the size of the data set?

Reply:
i guess because this scales in approximately the same way as partitioning the keyspace like i was trying to do before (the number of partitions is constant but the number of users per partition grows over time). we can rebalance i.e. tune the parameters as we see how it performs.

out of these three variables, one of them has to be dynamic:
1. chunk size
2. chunk interval
3. time to complete a full reconciliation cycle

i think we want #3 to be approximately constant as a policy decision so replication errors don't persist for longer than a day..

i don't think we're at risk of requests timing out any time soon due to the chunk size (which will be 4000 or less). and whether we make the chunk size or the chunk interval dynamic, the load on the accounts db is the same.

any way we slice it, we're going to end up monitoring and tuning whatever magic numbers we use.

Reply:
@acton-signal any thoughts?

Reply:
I think I would choose dynamic chunk interval and constant chunk size. This is so that you have a consistent database query and so that you have consistent memory usage. If the chunk grows larger in size, you would need a longer running query and a larger allocation of ram for usage. Not that it would blow up any time soon but it could eventually with a large enough data set.

Reply:
changed to status chunk size, dynamic interval in 238fc98d7912